pkg/controller/perconaservermongodb/psmdb_controller.go

```diff
@@ -387,49 +387,101 @@ func (r *ReconcilePerconaServerMongoDB) Reconcile(ctx context.Context, request r
 		}
 	}
 
-	shards := 0
-	for _, replset := range repls {
-		if (cr.Spec.Sharding.Enabled && replset.ClusterRole == api.ClusterRoleShardSvr) ||
-			!cr.Spec.Sharding.Enabled {
-			shards++
+	clusterStatus, err = r.reconcileReplsets(ctx, cr, repls)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "reconcile statefulsets")
+	}
+
+	err = r.stopMongosInCaseOfRestore(ctx, cr)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "on restore")
+	}
+
+	err = r.reconcileMongos(ctx, cr)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "reconcile mongos")
+	}
+
+	if err := r.enableBalancerIfNeeded(ctx, cr); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to start balancer")
+	}
+
+	if err := r.disableBalancerIfNeeded(ctx, cr); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to disable balancer")
+	}
+
+	if err := r.upgradeFCVIfNeeded(ctx, cr, *repls[0], cr.Status.MongoVersion); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to set FCV")
+	}
+
+	err = r.deleteMongosIfNeeded(ctx, cr)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "delete mongos")
+	}
+
+	err = r.deleteCfgIfNeeded(ctx, cr)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "delete config server")
+	}
+
+	// clean orphan PVCs if downscale
+	if isDownscale {
+		err = r.deleteOrphanPVCs(ctx, cr)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to delete orphan PVCs: %v", err)
 		}
+	}
+
+	err = r.exportServices(ctx, cr)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "export services")
+	}
+
+	err = r.scheduleEnsureVersion(ctx, cr, VersionServiceClient{})
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to ensure version")
+	}
+
+	if err = r.updatePITR(ctx, cr); err != nil {
+		return rr, err
+	}
+
+	err = r.resyncPBMIfNeeded(ctx, cr)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "resync PBM if needed")
+	}
+
+	return rr, nil
+}
 
+func (r *ReconcilePerconaServerMongoDB) reconcileReplsets(ctx context.Context, cr *api.PerconaServerMongoDB, repls []*api.ReplsetSpec) (api.AppState, error) {
+	log := logf.FromContext(ctx)
+
+	for _, replset := range repls {
 		if cr.Spec.Sharding.Enabled && replset.ClusterRole != api.ClusterRoleConfigSvr && replset.Name == api.ConfigReplSetName {
-			return reconcile.Result{}, errors.Errorf("%s is reserved name for config server replset", api.ConfigReplSetName)
+			return "", errors.Errorf("%s is reserved name for config server replset", api.ConfigReplSetName)
```

Comment:
Since you added `AppStateNone` why don't you return that rather than this empty string?

Reply:
This value is ignored when an error is returned. I think it's better to not use const instead of an empty string because if `AppStateNone` changes in the future, we would still want to have an empty string here